### ISSUE TITLE
datalad run  misses original files which were renamed and leaves them deleted

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#2276]
#### What steps will reproduce the problem?
```shell
$> datalad create /tmp/testrunrename                           
[INFO   ] Creating a new annex repo at /tmp/testrunrename 
create(ok): /tmp/testrunrename (dataset)                                                                                                                                                                                                      

$> cd /tmp/testrunrename

$> echo 123 > 123; datalad add 123
add(ok): /tmp/testrunrename/123 (file)                                                                                                                                                                                                        
save(ok): /tmp/testrunrename (dataset)                                                                                                                                                                                                        
action summary:
  add (ok: 1)
  save (ok: 1)

$> datalad run "git mv 123 124"
[INFO   ] == Command start (output follows) ===== 
[INFO   ] == Command exit (modification check follows) ===== 
add(notneeded): 124 (file) [already included in the dataset]
save(ok): /tmp/testrunrename (dataset)
action summary:
  add (notneeded: 1)
  save (ok: 1)

$> git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    123

```
#### What version of DataLad are you using (run `datalad --version`)? On what operating system (consider running `datalad plugin wtf`)?
0.9.2

**Comment by kyleam** [comment]
Hmm, I'm puzzled.  Why does `datalad add .` result in the deleted file from the `git mv` operation being commited, while the `ds.add` call in `run.py` doesn't?

Or as a quick test

```python
@with_tree({"file": "empty"})
def test_add_git_mv(path):
    ds = Dataset(path).create(force=True)
    ds.add(".")
    ok_clean_git(ds.path)

    from datalad.cmd import GitRunner
    GitRunner().run(["git", "mv", "file", "new"], cwd=path)
    ds.add(".", recursive=True)
    ok_clean_git(ds.path)
```

This will fail.  But if I go to the directory of the failed test and run "datalad add .", the deleted file is committed.

**Comment by bpoldrack** [comment]
I tried to reproduce via plain shell. It somehow depends on whether or not the dataset to add to is specified.

```
ben@tree /tmp % datalad create test-mv
[INFO   ] Creating a new annex repo at /tmp/test-mv 
create(ok): /tmp/test-mv (dataset)                                                                       
ben@tree /tmp % cd test-mv 
ben@tree /tmp/test-mv % touch some
ben@tree /tmp/test-mv % datalad add .
add(ok): /tmp/test-mv/some (file)                                                                        
add(ok): /tmp/test-mv (dataset)
save(ok): /tmp/test-mv (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)
ben@tree /tmp/test-mv % git status
On branch master
nothing to commit, working tree clean
ben@tree /tmp/test-mv % ll
total 380
drwxr-xr-x  4 ben  ben    4096 Mar  7 15:14 .
drwxrwxrwt 25 root root 364544 Mar  7 15:14 ..
drwxr-xr-x  2 ben  ben    4096 Mar  7 15:14 .datalad
drwxr-xr-x  9 ben  ben    4096 Mar  7 15:14 .git
-rw-r--r--  1 ben  ben      21 Mar  7 15:14 .gitattributes
lrwxrwxrwx  1 ben  ben     108 Mar  7 15:14 some -> .git/annex/objects/2W/kW/MD5E-s0--d41d8cd98f00b204e9800998ecf8427e/MD5E-s0--d41d8cd98f00b204e9800998ecf8427e
ben@tree /tmp/test-mv % git mv some new
ben@tree /tmp/test-mv % git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	renamed:    some -> new

ben@tree /tmp/test-mv % datalad add .
add(notneeded): /tmp/test-mv (dataset) [nothing to add from /tmp/test-mv]
save(ok): /tmp/test-mv (dataset)
action summary:
  add (notneeded: 1)
  save (ok: 1)
ben@tree /tmp/test-mv % git status
On branch master
nothing to commit, working tree clean
```
So far all good - that's what @kyleam was saying, right?
Now, do it again, but specify the dataset to add to explicitly:

```
ben@tree /tmp/test-mv % touch another
ben@tree /tmp/test-mv % datalad add .
add(ok): /tmp/test-mv/another (file)      
add(ok): /tmp/test-mv (dataset)
save(ok): /tmp/test-mv (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)
ben@tree /tmp/test-mv % ll
total 384
drwxr-xr-x  4 ben  ben    4096 Mar  7 15:16 .
drwxrwxrwt 25 root root 364544 Mar  7 15:16 ..
lrwxrwxrwx  1 ben  ben     108 Mar  7 15:15 another -> .git/annex/objects/2W/kW/MD5E-s0--d41d8cd98f00b204e9800998ecf8427e/MD5E-s0--d41d8cd98f00b204e9800998ecf8427e
drwxr-xr-x  2 ben  ben    4096 Mar  7 15:14 .datalad
drwxr-xr-x  9 ben  ben    4096 Mar  7 15:16 .git
-rw-r--r--  1 ben  ben      21 Mar  7 15:14 .gitattributes
lrwxrwxrwx  1 ben  ben     108 Mar  7 15:14 new -> .git/annex/objects/2W/kW/MD5E-s0--d41d8cd98f00b204e9800998ecf8427e/MD5E-s0--d41d8cd98f00b204e9800998ecf8427e
ben@tree /tmp/test-mv % git mv another brandnew
ben@tree /tmp/test-mv % git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	renamed:    another -> brandnew

ben@tree /tmp/test-mv % datalad add -d . .
add(notneeded): brandnew (file) [already included in the dataset]
save(ok): /tmp/test-mv (dataset)
action summary:
  add (notneeded: 1)
  save (ok: 1)
ben@tree /tmp/test-mv % git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    another
```

So, it went from a staged renaming to a staged deletion, which is why it isn't clean after the first call to add, but after the next one.

**Comment by bpoldrack** [comment]
Tried both approaches with debug output enabled.
In both cases the setup is:
```
datalad create test
cd test
touch some
datalad add .
git mv some new
```

Now, `datalad add -d. .` with debug:
```
% DATALAD_LOG_LEVEL=DEBUG datalad add -d . .
[DEBUG  ] parseParameters: Given "exclude_special_remotes: bool, optional
|   if True, don't return annex special remotes", we split into [('exclude_special_remotes', "exclude_special_remotes: bool, optional\n  if True, don't return annex special remotes")] 
[DEBUG  ] parseParameters: Given "with_urls_only : bool, optional
|   return only remotes which have urls", we split into [('with_urls_only', 'with_urls_only : bool, optional\n  return only remotes which have urls')] 
[DEBUG  ] Building doc for <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Building doc for <class 'datalad.interface.save.Save'> 
[DEBUG  ] Building doc for <class 'datalad.distribution.add.Add'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.add.Add'> 
[DEBUG  ] Building doc for <class 'datalad.plugin.Plugin'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Building doc for <class 'datalad.interface.diff.Diff'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Adding content to repo <AnnexRepo path=/tmp/test-mv (<class 'datalad.support.annexrepo.AnnexRepo'>)>: {'/tmp/test-mv/new': {'revision_src': 'ce380320e1ddbc7ecc482fa5f49e6e2bccb6eec6', 'path_src': '/tmp/test-mv/some', 'mode_src': 40960, 'state': 'renamed', 'perc_similarity': 100.0, 'mode': 40960, 'parentds': '/tmp/test-mv', 'action': 'add', 'path': '/tmp/test-mv/new', 'type_src': 'file', 'type': 'file', 'refds': '/tmp/test-mv', 'revision': 'ce380320e1ddbc7ecc482fa5f49e6e2bccb6eec6'}} 
[DEBUG  ] already included in the dataset [add(/tmp/test-mv/new)] 
add(notneeded): new (file) [already included in the dataset]
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.save.Save'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] already included in the dataset [save(/tmp/test-mv/new)] 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] already included in the dataset [save(/tmp/test-mv/new)] 
[DEBUG  ] Committing via direct call of git: ['git', 'commit', '-m', '[DATALAD] added content'] 
save(ok): /tmp/test-mv (dataset)
action summary:
  add (notneeded: 1)
  save (ok: 1)

% git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    some
```

and `datalad add .` instead:
```% DATALAD_LOG_LEVEL=DEBUG datalad add .              
[DEBUG  ] parseParameters: Given "exclude_special_remotes: bool, optional
|   if True, don't return annex special remotes", we split into [('exclude_special_remotes', "exclude_special_remotes: bool, optional\n  if True, don't return annex special remotes")] 
[DEBUG  ] parseParameters: Given "with_urls_only : bool, optional
|   return only remotes which have urls", we split into [('with_urls_only', 'with_urls_only : bool, optional\n  return only remotes which have urls')] 
[DEBUG  ] Building doc for <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Building doc for <class 'datalad.interface.save.Save'> 
[DEBUG  ] Building doc for <class 'datalad.distribution.add.Add'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.add.Add'> 
[DEBUG  ] Building doc for <class 'datalad.plugin.Plugin'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Adding content to repo <AnnexRepo path=/tmp/test-mv2 (<class 'datalad.support.annexrepo.AnnexRepo'>)>: {'/tmp/test-mv2': {'orig_request': '.', 'action': 'add', 'path': '/tmp/test-mv2', 'type': 'dataset', 'raw_input': True}} 
[DEBUG  ] nothing to add from /tmp/test-mv2 [add(/tmp/test-mv2)] 
add(notneeded): /tmp/test-mv2 (dataset) [nothing to add from /tmp/test-mv2]
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.save.Save'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] nothing to add from /tmp/test-mv2 [save(/tmp/test-mv2)] 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] nothing to add from /tmp/test-mv2 [save(/tmp/test-mv2)] 
[DEBUG  ] Committing via direct call of git: ['git', 'commit', '-m', '[DATALAD] added content'] 
save(ok): /tmp/test-mv2 (dataset)
action summary:
  add (notneeded: 1)
  save (ok: 1)

% git status
On branch master
nothing to commit, working tree clean
```

Clearly, the "Adding content to repo" is significantly different.


Re @yarikoptic:
Sure, but here it should result in the same thing. We are at the top of the dataset and the automatically determined dataset is the very same as the one specified. So, that doesn't explain the effect we observe.

**Comment by bpoldrack** [comment]
I guess, the difference it discovering what to add via AnnotatePath is okay in general, but what is incorrect therein is that we do not actually call annex-add correctly, since AnnotatePath is under the impression, that this is "notneeded", because the new file is "known" to annex. But this is only because annex does know that key already. So, in the case the discovery leads to both files explicitly, it gets added only "half-way". `results_from_annex_noinfo` should be the key piece, that needs to consider things differently if there is a renaming.

**Comment by yarikoptic** [comment]
@bpoldrack I do not think it is rooted in annex since happens to pure git repositories as well:
```shell
(git)hopa:/tmp/test[master]
$> builtin cd /tmp/; datalad remove --nocheck test; datalad create --no-annex test && cd test && touch some && datalad add . && git mv some new && datalad add -d . . && git status    
remove(ok): /tmp/test (dataset)
[INFO   ] Creating a new git repo at /tmp/test 
create(ok): /tmp/test (dataset)
add(ok): /tmp/test/some (file)
add(ok): /tmp/test (dataset)
save(ok): /tmp/test (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)
add(notneeded): new (file) [already included in the dataset]
save(ok): /tmp/test (dataset)
action summary:
  add (notneeded: 1)
  save (ok: 1)
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    some
```

**Comment by yarikoptic** [comment]
hm, I wonder if we could even call it a "feature" since "add" implies adding, not removing ;) 
that is why add/AnnotatePaths does not see the deleted file since it cannot be added.  `save` would correctly save the changes, so I am wondering if run should run `add` with `save=False` and then call `save` explicitly for that reason

**Comment by yarikoptic** [comment]
ha ha -- one more effect: adding `-r` to save  causes it also not to save the deleted file.
Without `-r` works fine:
```shell
$> builtin cd /tmp/; datalad remove --nocheck test; datalad create --no-annex test && cd test && touch some && datalad add . && git mv some new && datalad add --nosave -d . . && datalad save -d . -m "bu" && git status 
remove(ok): /tmp/test (dataset)
[INFO   ] Creating a new git repo at /tmp/test 
create(ok): /tmp/test (dataset)
add(ok): /tmp/test/some (file)
add(ok): /tmp/test (dataset)
save(ok): /tmp/test (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)
add(notneeded): new (file) [already included in the dataset]
save(ok): /tmp/test (dataset)
On branch master
nothing to commit, working tree clean
```
with `-r` the same story:
```shell
$> builtin cd /tmp/; datalad remove --nocheck test; datalad create --no-annex test && cd test && touch some && datalad add . && git mv some new && datalad add --nosave -d . . && datalad save -d . -r -m "bu" && git status 
remove(ok): /tmp/test (dataset)
[INFO   ] Creating a new git repo at /tmp/test 
create(ok): /tmp/test (dataset)
add(ok): /tmp/test/some (file)
add(ok): /tmp/test (dataset)
save(ok): /tmp/test (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)
add(notneeded): new (file) [already included in the dataset]
save(ok): /tmp/test (dataset)
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    some

cached/staged changes:                                                           
 some | 0
```

**Comment by kyleam** [comment]
> I wonder if we could even call it a "feature" since "add" implies adding, not removing ;)

So, if we went that direction, we'd have to 1) change `datalad add .` (no `-r`) to behave that way and 2) update add's help message because it says "all modifications to a dataset are immediately saved".

**Comment by yarikoptic** [comment]
I think I have a fix... yet to check if the `-d .` vs `` is still then there

**Comment by yarikoptic** [comment]
fixed by #2305 2276 (in 0.9.x)
